File: 2019-03-04-496,503,556/lb/496.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> nums   = new Stack<>();
        HashMap<Integer,Integer> ret = new HashMap<>();
        int[] result = new int[nums1.length];
        int temp;
        for (int i : nums2){
            if (nums.size()==0 || nums.peek()>=i){
                nums.push(i);
            }else {
                while(nums.size()>0 && nums.peek()<i){
                    ret.put(nums.pop(),i);
                }
                nums.push(i);
            }
        }
        while (nums.size()>0){
            ret.put(nums.pop(),-1);
        }
        for (int i=0;i<nums1.length;i++){
            result[i] = ret.get(nums1[i]);
        }
        return result;
    }
}

File: 2019-03-04-496,503,556/watercow/496.py
#!/usr/bin/env python  
#-*- coding:utf-8 _*-  
""" 
@author:watercow 
@license: Apache Licence 
@file: test.py 
@time: 2019/02/18
@site:  
@software: PyCharm 

# code is far away from bugs with the god animal protecting
    I love animals. They taste delicious.
              ┏┓      ┏┓
            ┏┛┻━━━┛┻┓
            ┃      ☃      ┃
            ┃  ┳┛  ┗┳  ┃
            ┃      ┻      ┃
            ┗━┓      ┏━┛
                ┃      ┗━━━┓
                ┃  神兽保佑    ┣┓
                ┃　永无BUG！   ┏┛
                ┗┓┓┏━┳┓┏┛
                  ┃┫┫  ┃┫┫
                  ┗┻┛  ┗┻┛ 
"""
import re
import pickle

# 使用单调栈，对nums2进行入栈操作
# 同时使用hashmap进行记录

class Solution(object):
    def nextGreaterElement(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        stack = []
        hashmap = {}

        for i, num in enumerate(nums2):
            while len(stack) and stack[-1] < num:
                hashmap[stack.pop()] = num

            stack.append(num)


        for i, num in enumerate(nums1):
            nums1[i] = hashmap.get(num, -1)

        return nums1